### PR TITLE
Add libircclient package.

### PR DESCRIPTION
**Pull request by sgielen**
Please check if I did this correctly :)

The package builds and the contents of the resulting distfiles seem correct. I haven't yet tested the library against an actual CloudABI application, that's my next TODO.

One thing I'm not sure about: the two patches have an overlapping hunk, with some context in between that stays the same. If applied in the wrong order, one of them will have partially invalid context. Is that a problem in your opinion?

```patch
--- src/libircclient.c
+++ src/libircclient.c
@@ -859,6 +946,10 @@ int irc_process_select_descriptors (irc_session_t * session, fd_set *in_set, fd_
                        fprintf (stderr, "[DEBUG] Detected local address: %s\n", inet_ntoa(session->local_addr));
 #endif
 
+#else /* defined(CLOUDABI) */
+               // Assume it's already connected
+#endif
+
                session->state = LIBIRC_STATE_CONNECTED;
 
                // Prepare the data, which should be sent to the server
```

```patch
--- src/libircclient.c
+++ src/libircclient.c
@@ -861,10 +861,6 @@ int irc_process_select_descriptors (irc_session_t * session, fd_set *in_set, fd_
 
                session->state = LIBIRC_STATE_CONNECTED;
 
-               // Get the hostname
-       if ( gethostname (hname, sizeof(hname)) < 0 )                                                                       
-               strcpy (hname, "unknown");
-
                // Prepare the data, which should be sent to the server
                if ( session->server_password )
                {
```

### CODE REVIEW

**Comment by sgielen** [comment]
FYI patches are also reported upstream, because I hope one of them will be merged:

https://sourceforge.net/p/libircclient/patches/7/
https://sourceforge.net/p/libircclient/patches/8/

File: packages/libircclient/BUILD [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# This file was written by Sjors Gielen. Copyright is transferred to Nuxi
```
Comment (EdSchouten) [inline packages/libircclient/BUILD]: Just stick your copyright on this, man!

File: packages/libircclient/patch-add-cloudabi [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+ 	AC_CHECK_LIB(crypto, [CRYPTO_new_ex_data], [], [AC_MSG_ERROR([OpenSSL not found])])
+ 	AC_CHECK_LIB(ssl,    [SSL_library_init], [], [AC_MSG_ERROR([OpenSSL not found])])
+@@ -111,5 +131,6 @@ AC_SUBST(CFLAGS)
+ AC_SUBST(LDFLAGS)
+ AC_SUBST(LIBS)
+ AC_SUBST(PREFIX)
++AC_SUBST(BUILD_EXAMPLES)
+ AC_CONFIG_FILES([examples/Makefile src/Makefile])
+ AC_OUTPUT
+--- examples/Makefile.in
```
Comment (EdSchouten) [inline packages/libircclient/patch-add-cloudabi]: Instead of doing this, you could also just remove `examples/` from the `SUBDIRS` variable, right?

File: packages/libircclient/patch-add-cloudabi [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+@@ -99,6 +109,16 @@ case "$host" in
+ 	*) AC_MSG_RESULT([no]) ;;
+ esac
+ 
++# Checking for cloudabi
++AC_MSG_CHECKING([for CloudABI])
++case "$host" in
++	*-cloudabi*)
++		AC_MSG_RESULT([yes])
++		CFLAGS="$CFLAGS -DCLOUDABI"
```
Comment (EdSchouten) [inline packages/libircclient/patch-add-cloudabi]: Is there a reason why you can't use `__CloudABI__`?

File: packages/libircclient/patch-fix-warnings [hunk 4]
```diff
@@ -0,0 +1,99 @@
+--- src/colors.c
```
Comment (EdSchouten) [inline packages/libircclient/patch-fix-warnings]: Instead of making these changes, it's also perfectly fine to add `unsafe_string_sources` to the `BUILD` file. Just grep other packages for examples. This puts a `#define` at the top of the source file that makes cloudlibc expose the unsafe string functions.

I agree it's better if upstream moves away from the unsafe string functions, but I'd rather not make such changes locally. If we do something wrong with porting these functions over to use the safe versions, we could in fact create security bugs that weren't there before...

File: packages/libircclient/patch-add-cloudabi (hunk 2 above)
Comment (EdSchouten) [reply packages/libircclient/patch-add-cloudabi]: Just do a grep for `SUBDIRS` on the tree. :-)

**Comment by EdSchouten** [comment]
With respect to the overlapping chunks: cloudabi-ports does indeed apply patches in random order. Scripts like `fixup_patches.py` also regenerate diffs on top of cleanly extracted source tarballs.

I'd rather stick to this approach, as having patches depend on each other is also less than ideal. It doesn't allow you to easily remove a single patch. For now, having both those changes in a single diff is perfectly fine!

File: packages/libircclient/patch-add-cloudabi (hunk 3 above)
Comment (sgielen) [reply packages/libircclient/patch-add-cloudabi]: I just thought it'd be nicer if the build system had some control over it. But then again, it would have been better to add HAS_CONNECT and HAS_SOCKET in that case. I guess it's better to use `__CloudABI__` now that I think about it.

File: packages/libircclient/patch-add-cloudabi (hunk 2 above)
Comment (sgielen) [reply packages/libircclient/patch-add-cloudabi]: No, annoyingly not, because the root Makefile is not generated, it's fixed. Only examples/Makefile is generated, so that's why I did it there. I could patch the root Makefile, but then it would be hard to remove examples/ conditionally.

File: packages/libircclient/patch-fix-warnings (hunk 4 above)
Comment (sgielen) [reply packages/libircclient/patch-fix-warnings]: Yeah, I preferred using the safe functions instead of the unsafe one, also in the hopes that upstream will apply this.

File: packages/libircclient/BUILD (hunk 1 above)
Comment (sgielen) [reply packages/libircclient/BUILD]: Meh. Allright, I don't really care ;)

File: packages/libircclient/patch-add-cloudabi (hunk 2 above)
Comment (EdSchouten) [reply packages/libircclient/patch-add-cloudabi]: Well, for sending upstream, you of course want to make this conditional. For the package we're creating here, it would make more sense to come up with the smallest patch needed to solve the job at hand. What about just doing this?

```
--- Makefile
+++ Makefile
@@ -1,6 +1,6 @@
 # $Id: Makefile 84 2009-01-10 01:37:29Z chromerium $
 
-SUBDIRS = src examples
+SUBDIRS = src
 
 all:
        for subdir in $(SUBDIRS); do \
```

File: packages/libircclient/patch-add-cloudabi (hunk 3 above)
Comment (EdSchouten) [reply packages/libircclient/patch-add-cloudabi]: Agreed. Either test for function presence with Autoconf or use `__CloudABI__`. Testing for operating system through Autoconf makes little sense.

**Comment by sgielen** [comment]
@EdSchouten commit updated: it's now a single patch, I use `__CloudABI__` now and there are some more small changes. One major change is I remove the check for a compatible `malloc`, as it failed (didn't investigate why) and decided to put a `#define malloc rpl_malloc` in the `config.h`, causing a linking failure when trying to use the library.

It's linking now, I'm testing it. :)

**Comment by EdSchouten** [comment]
The GNU libc compatible malloc thing breaks because of this:

```
{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for GNU libc compatible malloc" >&5
$as_echo_n "checking for GNU libc compatible malloc... " >&6; }                  
if ${ac_cv_func_malloc_0_nonnull+:} false; then :                                
  $as_echo_n "(cached) " >&6                                                     
else                                                                             
  if test "$cross_compiling" = yes; then :                                       
  ac_cv_func_malloc_0_nonnull=no                                                 
else                    
```

I suspect they want to have a `malloc()` that returns a non-NULL value when trying to allocate zero bytes, which our implementation also satisfies. I'd say that the most elegant way to deal with this is to remove any local changes and add `ac_cv_func_malloc_0_nonnull=yes` as a configure argument. We already do something like that for some other packages, like `packages/xz/BUILD`.

File: packages/libircclient/patch-add-cloudabi [hunk 5]
```diff
@@ -0,0 +1,406 @@
+--- Makefile
```
Comment (EdSchouten) [inline packages/libircclient/patch-add-cloudabi]: Please put this part in a separate file, e.g. `patch-no-examples`.

File: packages/libircclient/patch-add-cloudabi [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+@@ -1120,7 +1207,7 @@ int irc_cmd_notice (irc_session_t * session, const char * nch, const char * text
+ 
+ void irc_target_get_nick (const char * target, char *nick, size_t size)
+ {
+-	char *p = strstr (target, "!");
++	const char *p = strstr (target, "!");
+ 	unsigned int len;
+ 
+ 	if ( p )
+--- src/portable.c
```
Comment (EdSchouten) [inline packages/libircclient/patch-add-cloudabi]: This can also be placed in a separate patch file, e.g. `patch-strcasecmp`

File: packages/libircclient/patch-add-cloudabi [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+ 	ln -fs libircclient.so.$(APIVERSION) $(DESTDIR)@libdir@/libircclient.so
+ 
+--- src/colors.c
++++ src/colors.c
+@@ -50,7 +50,7 @@ static inline void libirc_colorparser_addorcat (char ** destline, unsigned int *
+ 
+ 	if ( *destline )
+ 	{
+-		strcpy (*destline, str);
++		strlcpy (*destline, str, *destlen);
```
Comment (EdSchouten) [inline packages/libircclient/patch-add-cloudabi]: To keep this package maintainable, I still suggest these changes are reverted and `unsafe_string_sources` is used instead.

File: packages/libircclient/patch-add-cloudabi [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
++	-mkdir -p $(DESTDIR)@prefix@/lib
+ 	-mkdir -p $(DESTDIR)@libdir@
+ 	cp ../include/libircclient.h $(DESTDIR)@prefix@/include/libircclient.h
+ 	cp ../include/libirc_errors.h $(DESTDIR)@prefix@/include/libirc_errors.h
+ 	cp ../include/libirc_events.h  $(DESTDIR)@prefix@/include/libirc_events.h 
+ 	cp ../include/libirc_options.h  $(DESTDIR)@prefix@/include/libirc_options.h 
+ 	cp ../include/libirc_rfcnumeric.h  $(DESTDIR)@prefix@/include/libirc_rfcnumeric.h 
++	cp libircclient.a $(DESTDIR)@prefix@/lib/libircclient.a
+ 	cp libircclient.so $(DESTDIR)@libdir@/libircclient.so.$(APIVERSION)
+ 	ln -fs libircclient.so.$(APIVERSION) $(DESTDIR)@libdir@/libircclient.so
```
Comment (EdSchouten) [inline packages/libircclient/patch-add-cloudabi:69]: Do these two commands work properly? As we're only building a static library, there is no way a `.so` can be installed, right?

Comment (sgielen) [reply packages/libircclient/patch-add-cloudabi:69]: True, but that error is ignored, so the copy just doesn't take place :)

**Comment by sgielen** [comment]
@EdSchouten patch is updated :-) This version of libircclient is tested in practice on Cosix.
